cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -852,24 +853,243 @@ struct LogbChecked {
+struct RoundImpl<T, RoundMode::UP> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::ceil(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::TOWARDS_ZERO> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::trunc(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::TOWARDS_INFINITY> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::signbit(val) ? std::floor(val) : std::ceil(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_DOWN> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return RoundImpl<T, RoundMode::DOWN>::Round(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_UP> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return RoundImpl<T, RoundMode::UP>::Round(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TOWARDS_ZERO> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return RoundImpl<T, RoundMode::TOWARDS_ZERO>::Round(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TOWARDS_INFINITY> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return RoundImpl<T, RoundMode::TOWARDS_INFINITY>::Round(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TO_EVEN> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::round(val * T(0.5)) * 2;
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TO_ODD> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::floor(val * T(0.5)) + std::ceil(val * T(0.5));
+  }
+};
+
+// Specializations of kernel state for round kernels
+template <typename>
+struct RoundOptionsWrapper;
+
+template <>
+struct RoundOptionsWrapper<RoundOptions> : public OptionsWrapper<RoundOptions> {
+  using OptionsType = RoundOptions;
+  double pow10;
+
+  explicit RoundOptionsWrapper(OptionsType options) : OptionsWrapper(std::move(options)) {
+    // Only positive of powers of 10 are used because combining multiply and
+    // division operations produced more stable rounding than using multiply-only.
+    // Refer to NumPy's round implementation:
+    // https://github.com/numpy/numpy/blob/7b2f20b406d27364c812f7a81a9c901afbd3600c/numpy/core/src/multiarray/calculation.c#L589
+    pow10 = RoundUtil::Pow10(std::abs(options.ndigits));
+  }
+
+  static Result<std::unique_ptr<KernelState>> Init(KernelContext* ctx,
+                                                   const KernelInitArgs& args) {
+    if (auto options = static_cast<const OptionsType*>(args.options)) {
+      return arrow::internal::make_unique<RoundOptionsWrapper<OptionsType>>(*options);
+    }
+
+    return Status::Invalid(
+        "Attempted to initialize KernelState from null FunctionOptions");
+  }
```

Comment:
Should this just return OptionsWrapper<OptionsType>::Init like below?

Reply:
No, its is `RoundOptionsWrapper` so that its constructor is invoked via `make_unique`, which initializes `pow10` data member. If we use `OptionsWrapper` then `pow10` is not available. I tried invoking `OptionsWrapper::Init` but it returns a `std::unique_ptr` which would require "casting" to `RoundOptionsWrapper` first and then to `KernelState` to match return type. The `unique_ptr` casting caused too many issues so I reverted to mimic the `OptionsWrapper::Init` method.

Reply:
The way I used these `OptionsWrapper` classes:
* considers `Init` for options validation because `Init` can return a `Status::Invalid`
* constructor for initializing non-options state, which can then be accessed in kernels' `Call` via `ctx->state()`

Reply:
I templetized `Init` method of `KernelState` derived classes so that derived constructors can be invoked without the need to duplicate the `Init` definition. This makes `KernelState` be fully functional to support options validation and extending kernel state via constructor which can most likely benefit other scalar kernels as well.

For example by extending, `OptionsWrapper` as follows:
```c++
template <typename OptionsType>
struct OptionsWrapper : public KernelState {
  template <typename KernelStateType = OptionsWrapper>
  static Result<std::unique_ptr<KernelState>> Init(KernelContext* ctx, const KernelInitArgs& args) {
    if (auto options = static_cast<const OptionsType*>(args.options)) {
        return ::arrow::internal::make_unique<KernelStateType>(*options);
    }
    ...
  }
  ...
};
```
now we can extend custom states as follows:
```c++
struct RoundOptionsWrapper<RoundOptions> : public OptionsWrapper<RoundOptions> {
  using OptionsType = RoundOptions;
  using State = RoundOptionsWrapper<OptionsType>;
  double pow10;

  explicit RoundOptionsWrapper(OptionsType options) : OptionsWrapper(std::move(options)) {
    pow10 = RoundUtil::Pow10(std::abs(options.ndigits));
  }

  static Result<std::unique_ptr<KernelState>> Init(KernelContext* ctx, const KernelInitArgs& args) {
    return OptionsWrapper<OptionsType>::Init<State>(ctx, args);
  }
};
```

Reply:
Ok, I tried the template variant of `OptionsWrapper` and although it made the code cleaner, it failed to compiled for some systems, so reverted to duplicating the `OptionsWrapper::Init` definition. I think there needs to be a refactoring of KernelState and related parts to support validating kernel options and extending kernel state in a simpler manner. There are different patterns being used in the code to fulfill these. But this is a separate issue from this PR.